File: java/HouseRobber.java
import java.util.Arrays;

public class HouseRobber {
    public int rob(int[] nums) {
        int valueOfAdjacent = nums[nums.length - 1];
        int valueOfNextAdjacent = 0;
        int value = 0;
        for(int i = nums.length - 2; i >= 0; i--) {
            value = Math.max(nums[i] + valueOfNextAdjacent, valueOfAdjacent);

            valueOfNextAdjacent = valueOfAdjacent;
            valueOfAdjacent = value;
        }

        return Math.max(value, valueOfAdjacent);
    }

    public static void main(String[] args) {
        HouseRobber hr = new HouseRobber();
        System.out.println(hr.rob(new int[]{1, 2, 3, 1}));
        System.out.println(hr.rob(new int[]{2, 7, 9, 3, 1}));
        System.out.println(hr.rob(new int[]{1}));
        System.out.println(hr.rob(new int[]{1, 3}));
        System.out.println(hr.rob(new int[]{ 3, 1, 5}));
    }
}

File: python/house_robber.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        if len(nums) <= 2:
            return max(nums[-1], nums[-2])

        nums[2] = max(nums[0] + nums[2], nums[1])
        for index in range(3, len(nums)):
            nums[index] = max(nums[index - 1], nums[index] + nums[index - 2], nums[index] + nums[index - 3])

        return max(nums[-1], nums[-2])


s = Solution()
print(s.rob([1,2,3,1]))
print(s.rob([2,7,9,3,1]))
print(s.rob([1,2]))
print(s.rob([1,]))
print(s.rob([2,1,1,2]))
print(s.rob([1,7,9,2]))